dipy/reconst/ivim.py

```diff
@@ -0,0 +1,487 @@
+from __future__ import division, print_function, absolute_import
+
+from distutils.version import LooseVersion
+
+import numpy as np
+import scipy
+import warnings
+from dipy.reconst.base import ReconstModel
+from dipy.reconst.multi_voxel import multi_voxel_fit
+
+SCIPY_LESS_0_17 = LooseVersion(scipy.version.short_version) < '0.17'
+
+if SCIPY_LESS_0_17:
+    from scipy.optimize import leastsq
+else:
+    from scipy.optimize import least_squares
+
+
+def ivim_prediction(params, gtab, S0=1.):
+    """The Intravoxel incoherent motion (IVIM) model function.
+
+    Parameters
+    ----------
+    params : array
+        An array of IVIM parameters - [S0, f, D_star, D].
+
+    gtab : GradientTable class instance
+        Gradient directions and bvalues.
+
+    S0 : float, optional
+        This has been added just for consistency with the existing
+        API. Unlike other models, IVIM predicts S0 and this is over written
+        by the S0 value in params.
+    """
+    S0, f, D_star, D = params
+    b = gtab.bvals
+    S = S0 * (f * np.exp(-b * D_star) + (1 - f) * np.exp(-b * D))
+    return S
+
+
+def _ivim_error(params, gtab, signal):
+    """Error function to be used in fitting the IVIM model.
+
+    Parameters
+    ----------
+    params : array
+        An array of IVIM parameters - [S0, f, D_star, D]
+
+    gtab : GradientTable class instance
+        Gradient directions and bvalues.
+
+    signal : array
+        Array containing the actual signal values.
+    """
+    return signal - ivim_prediction(params, gtab)
+
+
+def f_D_star_prediction(params, gtab, x0):
+    """Function used to predict IVIM signal when S0 and D are known
+    by considering f and D_star as the unknown parameters.
+
+    Parameters
+    ----------
+    params : array, dtype=float
+        An array containing the values of f and D_star.
+
+    gtab : GradientTable class instance
+        Gradient directions and bvalues.
+
+    x0 : array, dtype=float
+        The parameters - [S0, f, D_star, D] obtained by a linear fit.
+    """
+    f, D_star = params
+    S0, D = x0[0], x0[3]
+    b = gtab.bvals
+    S = S0 * (f * np.exp(-b * D_star) + (1 - f) * np.exp(-b * D))
+    return S
+
+
+def f_D_star_error(params, gtab, signal, x0):
+    """Error function used to fit f and D_star keeping S0 and D fixed
+
+    Parameters
+    ----------
+    params : array
+        The value of f and D_star.
+
+    gtab : GradientTable class instance
+        Gradient directions and bvalues.
+
+    signal : array
+        Array containing the actual signal values.
+
+    x0 : array, dtype=float
+        The parameters - [S0, f, D_star, D] obtained by a linear fit.
+    """
+    f, D_star = params
+    return signal - f_D_star_prediction([f, D_star], gtab, x0)
+
+
+class IvimModel(ReconstModel):
+    """Ivim model
+    """
+
+    def __init__(self, gtab, split_b_D=400.0, split_b_S0=200., bounds=None,
+                 tol=1e-15, f_threshold=0.25, x_scale=np.array([1e03, 1e-01, 1e-03, 1e-4]),
```

Comment:
Would you mind writing the x_scale as : `[1000, 0.1, 0.001, 0.0001]` instead? It's easier to read, I think.